File: src-tauri/src/models/backup.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub enum BackupType {
    Full,
    Incremental,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BackupMeta {
    pub id: String,
    pub server_id: String,
    pub name: String,
    pub backup_type: BackupType,
    pub created_at: u64,
    pub size_bytes: u64,
    pub file_count: u32,
    pub parent_id: Option<String>,
    pub paths_included: Vec<String>,
    pub notes: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BackupFileEntry {
    pub relative_path: String,
    pub sha256: String,
    pub size: u64,
    pub modified_at: u64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BackupManifest {
    pub meta: BackupMeta,
    pub files: Vec<BackupFileEntry>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CreateBackupRequest {
    pub server_id: String,
    pub name: Option<String>,
    pub incremental: bool,
    pub include_paths: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RestoreBackupRequest {
    pub server_id: String,
    pub backup_id: String,
    pub overwrite: bool,
}

File: src/api/backup.ts
import { tauriInvoke } from "@api/tauri";

export interface BackupMeta {
  id: string;
  server_id: string;
  name: string;
  backup_type: "Full" | "Incremental";
  created_at: number;
  size_bytes: number;
  file_count: number;
  parent_id: string | null;
  paths_included: string[];
  notes: string | null;
}

export interface BackupFileEntry {
  relative_path: string;
  sha256: string;
  size: number;
  modified_at: number;
}

export interface BackupManifest {
  meta: BackupMeta;
  files: BackupFileEntry[];
}

export const backupApi = {
  async createBackup(
    serverId: string,
    name: string | null,
    incremental: boolean,
    includePaths: string[],
  ): Promise<BackupMeta> {
    return tauriInvoke("create_backup", {
      serverId,
      name,
      incremental,
      includePaths,
    });
  },

  async listBackups(serverId: string): Promise<BackupMeta[]> {
    return tauriInvoke("list_backups", { serverId });
  },

  async getBackupDetail(serverId: string, backupId: string): Promise<BackupManifest> {
    return tauriInvoke("get_backup_detail", { serverId, backupId });
  },

  async restoreBackup(serverId: string, backupId: string): Promise<void> {
    return tauriInvoke("restore_backup", { serverId, backupId });
  },

  async deleteBackup(serverId: string, backupId: string): Promise<void> {
    return tauriInvoke("delete_backup", { serverId, backupId });
  },
};
